docs/proposed/psa-driver-interface.md

```diff
@@ -646,6 +646,36 @@ The core will not update the persistent state in storage while an entry point is
 
 In a multithreaded environment, the driver may only call these two functions from the thread that is executing the entry point.
 
+#### Built-in keys
+
+Opaque drivers may declare built-in keys. Built-in keys can be accessed, but not created, through the PSA Cryptography API.
+
+A built-in key is identified by its location and its **slot number**. Drivers that support built-in keys must provide a `"get_builtin_key"` entry point to retrieve the key data and metadata. The core calls this entry point when it needs to access the key, typically because the application requested an operation on the key. The core may keep information about the key in cache, and successive calls to access the same slot number should return the same data. This entry point has the following prototype:
+
+```
+psa_status_t acme_get_builtin_key(psa_drv_slot_number_t slot_number,
+                                  psa_key_attributes_t *attributes,
+                                  uint8_t *key_buffer,
+                                  size_t key_buffer_size);
+```
+
+If this function returns `PSA_SUCCESS` or `PSA_ERROR_BUFFER_TOO_SMALL`, it must fill `attributes` with the attributes of the key (except for the key identifier). On success, this function must also fill `key_buffer` with the key context.
+
+On entry, `psa_get_key_lifetime(attributes)` is the location at which the driver was declared and the persistence level `#PSA_KEY_LIFETIME_PERSISTENT`. The driver entry point may change the lifetime to one with the same location but a different persistence level. The standard attributes other than the key identifier and lifetime have the value conveyed by `PSA_KEY_ATTRIBUTES_INIT`.
```

Comment:
I think we already discussed about that but why would you pass the application key identifier to the driver?

Reply:
Er, we don't?

> except for the key identifier

> attributes other than the key identifier and lifetime

The section “Key management in a secure element with storage” discusses why drivers shouldn't rely on the key identifier.

Reply:
I guess my question should rather had been: why on entry the value of the key identifier in attributes is not just the value conveyed by `PSA_KEY_ATTRIBUTES_INIT`?

Reply:
Typically the core would already have an attribute structure at that point with the key id specified by the application. Forcing the core to store the key id elsewhere until the driver has finished its work would make the core use a little more code and memory and the only advantage is to completely prevent dodgy drivers from peeking at the key attribute. There may even be specific implementations where it's fine for the driver to peek at key attributes, when the driver is only intended to work on specific single-application systems (though I'll agree that it's dodgy, because it makes it harder to later move to multi-application support).

Reply:
The driver don't need it thus I'd say just don't pass it. It's just going to be a source of problems to me.

Reply:
I forgot that the fact that we don't hide the key identifier in the key attributes is widespread in the driver interface. That way the PSA implementation doesn't have to patch key attributes passed by the application before to pass them to a driver. Thus no need to make this effort here either.